sys/sys/param.h

```diff
@@ -367,4 +367,12 @@ __END_DECLS
 #define __PAST_END(array, offset)	\
     (((__typeof__(*(array)) *)__builtin_no_change_bounds(array))[offset])
 
+#ifdef _KERNEL
+/*
+ * Convenience wrapper to convert an in-kernel pointer to a DDC-derived
+ * capability.  NB: For purecap kernels this is a no-op.
+ */
+#define	PTR2CAP(p)	((__cheri_tocap __typeof__((p)) __capability)(p))
```

Comment:
I somewhat wonder if we want to add an assertion that the capability is in the kernel address space.

Reply:
Bound DDC accordingly and let it trap automatically :)

Reply:
I would like to to bound DDC for the RISC-V hybrid kernel.  I think that should work at this point.  I might try it in a separate change though.  I wonder if we can get MIPS to work with that, too.

Reply:
The main issue that I can think of is C{To,From}Ptr being offset-based, so things may get confused and think that a kernel address is a user address if they end up using those... :/

Reply:
Oh, crud, __cheri_tocap doesn't use csetaddr but csetoffset (in effect)?  That means the kernel has to be linked at address 0 (well, offset by VM_MIN_KERNEL effectively) instead of KERNBASE and probably breaks all sorts of things.

Reply:
It has to, that's how DDC works; if the base (well, cursor) is non-zero then all your pointers need to be adjusted accordingly, because DDC gets added, it's not just the authorisation.

Reply:
I think I will add the assertion here then and punt on DDC trimming for hybrid.